File: Sources/DocShot/Models/RecordingRegionPlan.swift
import Foundation
import CoreGraphics

/// Decides whether a selected region can be recorded, and how it maps onto one display's
/// ScreenCaptureKit source rectangle.
///
/// R1 records a region only when it lies wholly within a single display. Anything else is
/// rejected with a reason the UI shows, because the alternatives — cropping to one display or
/// stretching a composite — would record something the user did not select.
///
/// The output pixel size is `points × that display's scale`, rounded to the nearest whole pixel
/// and otherwise preserved exactly. Odd dimensions are kept: the selected size is the promise.
public enum RecordingRegionPlan: Equatable, Sendable {
    case ok(displayID: CGDirectDisplayID, sourceRectInPoints: CGRect, outputPixelSize: CGSize)
    case rejected(RecordingTargetRejection)

    /// Smallest region side R1 accepts, in points.
    public static let minimumSideInPoints: CGFloat = 16

    /// - Parameters:
    ///   - globalRectInCG: the selection in global Core Graphics space (origin top-left of the
    ///     main display, Y down), as produced by `DisplayGeometry.cocoaToCGRect`.
    ///   - displays: every connected display, at its own backing scale.
    public static func make(
        globalRectInCG rect: CGRect,
        displays: [DisplayDescriptor]
    ) -> RecordingRegionPlan {
        let normalized = DisplayGeometry.normalizeRect(rect)

        guard normalized.origin.x.isFinite, normalized.origin.y.isFinite,
              normalized.width.isFinite, normalized.height.isFinite,
              normalized.width > 0, normalized.height > 0 else {
            return .rejected(.invalidGeometry)
        }

        guard normalized.width >= minimumSideInPoints, normalized.height >= minimumSideInPoints else {
            return .rejected(.tooSmall)
        }

        let usableDisplays = displays.filter(\.isValid)

        // Mirrored displays can share a frame. Lowest display ID wins so the plan is deterministic.
        let containing = usableDisplays
            .filter { $0.frameInCG.contains(normalized) }
            .sorted { $0.displayID < $1.displayID }

        guard let display = containing.first else {
            let overlapping = usableDisplays.filter { !$0.frameInCG.intersection(normalized).isNull }
            return .rejected(overlapping.count > 1 ? .spansMultipleDisplays : .noContainingDisplay)
        }

        let sourceRect = CGRect(
            x: normalized.origin.x - display.frameInCG.origin.x,
            y: normalized.origin.y - display.frameInCG.origin.y,
            width: normalized.width,
            height: normalized.height
        )

        let pixelWidth = (normalized.width * display.scale).rounded()
        let pixelHeight = (normalized.height * display.scale).rounded()

        guard pixelWidth >= 1, pixelHeight >= 1 else {
            return .rejected(.tooSmall)
        }

        return .ok(
            displayID: display.displayID,
            sourceRectInPoints: sourceRect,
            outputPixelSize: CGSize(width: pixelWidth, height: pixelHeight)
        )
    }

    /// The recording target this plan describes, or `nil` when the selection was rejected.
    public var target: RecordingTarget? {
        guard case .ok(let displayID, let sourceRect, let outputSize) = self else { return nil }
        return .region(displayID: displayID, rectInDisplay: sourceRect, outputSize: outputSize)
    }

    public var rejection: RecordingTargetRejection? {
        guard case .rejected(let reason) = self else { return nil }
        return reason
    }
}

File: windows/src/DocShot.Core/Models/RecordingRegionPlan.cs
using DocShot.Core.Primitives;

namespace DocShot.Core.Models;

/// <summary>
/// Decides whether a selected region can be recorded, and how it maps onto one display's capture
/// source rectangle.
/// </summary>
/// <remarks>
/// A region records only when it lies wholly within a single display. Anything else is rejected
/// with a reason the UI shows, because the alternatives - cropping to one display or stretching a
/// composite - would record something the user did not select. The output pixel size is
/// <c>points x that display's scale</c>, rounded to the nearest whole pixel and otherwise
/// preserved exactly; odd dimensions are kept, matching the macOS R1 decision not to round them
/// away.
/// </remarks>
public abstract record RecordingRegionPlan
{
    public sealed record Ok(uint DisplayId, RectD SourceRectInPoints, SizeD OutputPixelSize) : RecordingRegionPlan;
    public sealed record Rejected(RecordingTargetRejection Reason) : RecordingRegionPlan;

    private RecordingRegionPlan() { }

    /// <summary>Smallest region side accepted, in points.</summary>
    public const double MinimumSideInPoints = 16;

    /// <param name="rect">The selection in global screen space (origin top-left, Y down).</param>
    /// <param name="displays">Every connected display, at its own scale.</param>
    public static RecordingRegionPlan Make(RectD rect, IReadOnlyList<DisplayDescriptor> displays)
    {
        var normalized = DisplayGeometry.NormalizeRect(rect);

        if (!double.IsFinite(normalized.X) || !double.IsFinite(normalized.Y) ||
            !double.IsFinite(normalized.Width) || !double.IsFinite(normalized.Height) ||
            normalized.Width <= 0 || normalized.Height <= 0)
        {
            return new Rejected(new RecordingTargetRejection.InvalidGeometry());
        }

        if (normalized.Width < MinimumSideInPoints || normalized.Height < MinimumSideInPoints)
        {
            return new Rejected(new RecordingTargetRejection.TooSmall());
        }

        var usableDisplays = displays.Where(d => d.IsValid).ToList();

        // Mirrored displays can share a frame. Lowest display ID wins so the plan is deterministic.
        var containing = usableDisplays
            .Where(d => d.FrameInScreen.Contains(normalized))
            .OrderBy(d => d.DisplayId)
            .ToList();

        if (containing.Count == 0)
        {
            var overlapping = usableDisplays.Where(d => d.FrameInScreen.Intersects(normalized)).ToList();
            return new Rejected(overlapping.Count > 1
                ? new RecordingTargetRejection.SpansMultipleDisplays()
                : new RecordingTargetRejection.NoContainingDisplay());
        }

        var display = containing[0];
        var sourceRect = new RectD(
            normalized.X - display.FrameInScreen.X,
            normalized.Y - display.FrameInScreen.Y,
            normalized.Width,
            normalized.Height);

        var pixelWidth = Math.Round(normalized.Width * display.Scale);
        var pixelHeight = Math.Round(normalized.Height * display.Scale);

        if (pixelWidth < 1 || pixelHeight < 1)
        {
            return new Rejected(new RecordingTargetRejection.TooSmall());
        }

        return new Ok(display.DisplayId, sourceRect, new SizeD(pixelWidth, pixelHeight));
    }

    /// <summary>The recording target this plan describes, or <c>null</c> when the selection was rejected.</summary>
    public RecordingTarget? Target => this is Ok ok
        ? new RecordingTarget.Region(ok.DisplayId, ok.SourceRectInPoints, ok.OutputPixelSize)
        : null;

    public RecordingTargetRejection? RejectionReason => this is Rejected r ? r.Reason : null;
}
